desc/compute/_neoclassical.py

```diff
@@ -0,0 +1,275 @@
+from .data_index import register_compute_fun
+from .utils import safediv
+
+
+def _vec_quadax(quad, **kwargs):
+    """Vectorize an adaptive quadrature method from quadax.
+
+    Parameters
+    ----------
+    quad : callable
+        Adaptive quadrature method matching API from quadax.
+
+    Returns
+    -------
+    vec_quad : callable
+        Vectorized adaptive quadrature method.
+
+    """
+
+    def vec_quad(fun, interval, B_sup_z, B, B_z_ra, arg1, arg2):
+        return quad(fun, interval, args=(B_sup_z, B, B_z_ra, arg1, arg2), **kwargs)[0]
+
+    vec_quad = jnp.vectorize(
+        vec_quad, signature="(2),(m),(m),(m),(m),(m)->()", excluded={0}
+    )
+    return vec_quad
+
+
+def _get_pitch(grid, min_B, max_B, num, for_adaptive=False):
+    """Get points for quadrature over velocity coordinate.
+
+    Parameters
+    ----------
+    grid : Grid
+        The grid on which data is computed.
+    min_B : jnp.ndarray
+        Minimum |B| value.
+    max_B : jnp.ndarray
+        Maximum |B| value.
+    num : int
+        Number of values to uniformly space in between.
+    for_adaptive : bool
+        Whether to return just the points useful for an adaptive quadrature.
+
+    Returns
+    -------
+    pitch : Array
+        Pitch values in the desired shape to use in compute methods.
+
+    """
+    min_B = grid.compress(min_B)
+    max_B = grid.compress(max_B)
+    if for_adaptive:
+        pitch = jnp.reciprocal(jnp.stack([max_B, min_B], axis=-1))[:, jnp.newaxis]
+        assert pitch.shape == (grid.num_rho, 1, 2)
+    else:
+        pitch = get_pitch(min_B, max_B, num)
+        pitch = jnp.broadcast_to(
+            pitch[..., jnp.newaxis], (pitch.shape[0], grid.num_rho, grid.num_alpha)
+        ).reshape(pitch.shape[0], grid.num_rho * grid.num_alpha)
+    return pitch
+
+
+def _poloidal_mean(grid, f):
+    """Integrate f over poloidal angle and divide by 2π."""
+    assert f.shape[-1] == grid.num_poloidal
+    if grid.num_poloidal == 1:
+        return jnp.squeeze(f, axis=-1)
+    if not hasattr(grid, "spacing"):
+        return jnp.mean(f, axis=-1)
+    assert grid.is_meshgrid
+    dp = grid.compress(grid.spacing[:, 1], surface_label="poloidal")
+    return f @ dp / jnp.sum(dp)
+
+
+@register_compute_fun(
+    name="<L|r,a>",
+    label="\\int_{\\zeta_{\\mathrm{min}}}^{\\zeta_{\\mathrm{max}}}"
+    " \\frac{d\\zeta}{|B^{\\zeta}|}",
+    units="m / T",
+    units_long="Meter / tesla",
+    description="(Mean) proper length of field line(s)",
+    dim=1,
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="r",
+    data=["B^zeta"],
+    resolution_requirement="z",  # and poloidal if near rational surfaces
+    source_grid_requirement={"coordinates": "raz", "is_meshgrid": True},
+)
+def _L_ra_fsa(data, transforms, profiles, **kwargs):
+    g = transforms["grid"].source_grid
+    shape = (g.num_rho, g.num_alpha, g.num_zeta)
+    L_ra = simpson(
+        y=jnp.reciprocal(data["B^zeta"]).reshape(shape),
```

Comment:
I will leave a series of comments like this, just asking for clarification, don't judge me if they are simple questions.

proper length I assumed had to be dl/|B| , what allows you to change to dzeta/ B^zeta?

Reply:
dl/B dot grad l = dzeta/B dot grad zeta = dtheta/B dot grad theta along a field line is the streamline condition.

Comes from dr cross B = 0 where dr is a infinitesimal vector element pointing along B.
dr  = partial r/partial theta dtheta + partial r/partial phi  dphi = e_theta dtheta + e_phi dphi.
Take cross with B = B^theta e_theta + B^phi e_phi and you get the streamline conditions.